sockeye/lexical_constraints.py

```diff
@@ -25,6 +25,104 @@
 RawConstraintList = List[List[int]]
 
 
+class AvoidPhrase:
+    """
+    Represents a negative phrasal constraint. Tracks progress through a phrase
+    and returns a word ID if that word ID would complete the phrase.
+
+    :param avoid_list: A sequence of integers in the target language vocabulary.
+    :param last_consumed: The last consumed word in the current hypothesis.
+    """
+    def __init__(self,
+                 avoid_list: List[int],
+                 last_consumed: int = -1) -> None:
+        self.phrase = avoid_list
+        self.last_consumed = last_consumed
+
+    def __str__(self):
+        return ' '.join(['*{}*'.format(x) if i == self.last_consumed else str(x) for i, x in enumerate(self.phrase)])
+
+    def consume(self, word_id: int) -> 'AvoidPhrase':
+        """
+        Consumes a word, and updates the tracking based on it. If the word is the next words in the phrase,
+        update last_consumed, otherwise, reset.
+
+        :param word_id: The word that was just generated.
+        """
+        if self.phrase[self.last_consumed + 1] == word_id:
+            return AvoidPhrase(self.phrase, self.last_consumed + 1)
```

Comment:
Do you need to return a new object? Couldn't you just update your last_consumed index as in the else: case? If so, you wouldn't actually need to return anything and the loop in AvoidBatch would be more efficient.

Reply:
AvoidPhrase objects can be shared so long as they have the same state. This is accomplished by returning a new object every time the state advances. This way every hypothesis has its own record of its progress through each phrase.

Put another way, the AvoidPhrase object is a reference shared across hypotheses in the beam. When any hypothesis advances, it gets a new version of the object that is specific to itself.